keps/sig-cli/20190805-kubectl-debug.md

```diff
@@ -0,0 +1,460 @@
+   via the Kubernetes API is acceptable.)
+
+New `kubectl` concepts increase cognitive burden for all users of Kubernetes.
+This KEP seeks to minimize this burden by mirroring the existing `kubectl`
+workflows where possible.
+
+### Non-Goals
+
+Ephemeral containers are supported on Windows, but it's not the recommended
+debugging facility. The other troubleshooting workflows described here are
+equally as useful to Windows containers. We will not attempt to create
+separate debugging facilities for Windows containers.
+
+[distroless container image]: https://github.com/GoogleCloudPlatform/distroless
+[distroless for k8s system images]: https://git.k8s.io/enhancements/keps/sig-release/20190316-rebase-images-to-distroless.md
+[scratch debugger]: https://github.com/kubernetes-retired/contrib/tree/master/scratch-debugger
+
+## Proposal
+
+### Pod Troubleshooting with Ephemeral Debug Container
+
+We will add a new command `kubectl debug` that will:
+
+1. Construct a `v1.Container` for the debug container based on command line
+   arguments. This will include an optional container for namespace targeting
+   (described below).
+2. Fetch the specified pod's existing ephemeral containers using
+   `GetEphemeralContainers()` in the generated pod client.
+3. Append the new debug container to the pod's ephemeral containers and call
+   `UpdateEphemeralContainers()`.
+4. Watch pod for updates to the debug container's `ContainerStatus` and
+   automatically attach once the container is running. (optional based on
+   command line flag)
+
+#### Debug Container Naming
+
+Currently, there is no support for deleting or recreating ephemeral containers.
+In cases where the user does not specify a name, `kubectl` should generate a
+unique name and display it to the user.
+
+### Container Namespace Targeting
+
+In order to see processes running in other containers in the pod, [Process
+Namespace Sharing] should be enabled for the pod. In cases where process
+namespace sharing isn't enabled for the pod, `kubectl` will set
+`TargetContainer` in the `EphemeralContainer`. This will cause the ephemeral
+container to be created in the namespaces of the target container in runtimes
+that support container namespace targeting.
+
+[Process Namespace Sharing]: https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace
+
+#### Interactive Troubleshooting and Automatic Attaching
+
+Since the primary use case for `kubectl debug` is interactive troubleshooting,
+`kubectl debug` will automatically attach to the console of the newly created
+ephemeral container and will default to creating containers with `Stdin` and
+`TTY` enabled.
+
+These may be disabled via command line flags.
+
+#### Proposed `kubectl debug` arguments
+
+```
+% kubectl help debug
+Execute a container in a pod.
+
+Examples:
+  # Get network address configuration from pod mypod
+  kubectl debug mypod ip addr list
```

Comment:
Given the image is required, this doesn't make sense, it's rather `kubectl exec ip addr list`, unless this is to run a command in the ephemeral container b/c the main is a distroless so we can't exec into it, in which case you'll probably want to add `--attach=false` too.

Reply:
good point, I neglected to update example when making image required. I would like to be able to set a default container image for debug, but I don't want to get hung up on it for v1.

`--attach` must be true because that's where the output of `ip addr list` will go. Unless we were to create the Ephemeral Container and do an exec rather than setting `Container.Args` to *args*. There would be some advantages to using exec this way, actually.

Rather than getting hung up on this, I've removed this example to focus on the interactive debugging use case for the first version.